pkg/virtual/framework/dynamic/apidefs/types.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apidefs
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	apiresourcev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+)
+
+// APIDomainKeyContextKeyType is the type of the key for the request context value
+// that will carry the API domain key.
+type APIDomainKeyContextKeyType string
+
+// APIDomainKeyContextKey is the key for the request context value
+// that will carry the API domain key.
+const APIDomainKeyContextKey APIDomainKeyContextKeyType = "VirtualWorkspaceAPIDomainLocationKey"
```

Comment:
APIDomainLocation is hard to understand. What is a location, what a domain?

Also, please expose `WithFoo(ctx, ...)` and `FooFrom(ctx)`  instead of the keys.

Why is this here in apidefs? Why does a context have something an APIDefinition or vice-versa?